staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+// For devices that are not local to a node, the node name is not set. Instead,
+// the driver may use a node selector to specify where the devices are available.
+type ResourceSlice struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Contains the information published by the driver.
+	//
+	// Changing the spec bumps up the generation number.
+	Spec ResourceSliceSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+
+	// Future extension: status.
+}
+
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// +required
+	DriverName string `json:"driverName" protobuf:"bytes,1,name=driverName"`
+
+	// PoolName is used to identify devices. For node-local devices, this
+	// is often the node name, but this is not required.
+	//
+	// It must not be longer than 253 and must consist of one or more DNS sub-domains
+	// separated by slashes.
+	//
+	// +required
+	PoolName string `json:"poolName" protobuf:"bytes,2,name=poolName"`
+
+	// NodeName identifies the node which provides the devices.
+	// A field selector can be used to list only ResourceSlice
+	// objects belonging to a certain node.
+	//
+	// This field can be used to limit access from nodes to slices with
+	// the same node name. It also indicates to autoscalers that adding
+	// new nodes of the same type as some old node might also make new
+	// devices available.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeName *string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+
+	// Defines which nodes have access to the devices in the pool.
+	// If the node selector is empty, all nodes have access.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,4,opt,name=nodeSelector"`
+
+	// The generation gets bumped in all slices of a pool whenever device
+	// definitions change. A consumer must only use device definitions from slices
+	// with the highest generation number and ignore all others.
+	PoolGeneration int64 `json:"poolGeneration" protobuf:"bytes,5,name=poolGeneration"`
+
+	// The total number of slices in the pool.
+	// Consumers can use this to check whether they have
+	// seen all slices.
+	PoolSliceCount int64 `json:"poolSliceCount" protobuf:"bytes,6,name=poolSliceCount"`
+
+	// Devices lists all available devices in this pool.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +required
```

Comment:
This needs to be `+optional` because
- a future alternative might get added (https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1655660015)
- a driver might want to advertise that it really has no devices on a node